### PR TITLE
Various bugfixes for cropping & rotation handling

### PR DESCRIPTION
**Pull request by stefan-it**
Hi,

this PR fixes various cropping & rotation handling bugs:

* Calculation of image boundary exceeding was done wrong: when e.g. image is rotated 90 or 270 degrees, the image dimensions switch. Refactored image boundary exceeding check was added to work with rotation at 90 and 270 degrees.
* In some cases, negative coordinates for region width and height could occur
* Also in some cases, region width and height could be less than the user-defined values. This was due to the fact, that `Math.min` was used. A new function was added that checks, if the calculates value for region width and height is >= the user-defined value.
* An extensive unit test was added, that uses a sliding window approach: a rectangle slides over an input image and rotation operations (90, 180 and 270 degress) are performed. The returned image is checked to match the user-defined dimensions. This check shows that no regression bugs were introduced with the new rotation handling and the library is now more robust when using rotation operations.

### CODE REVIEW

File: imageio-turbojpeg/src/main/java/de/digitalcollections/turbojpeg/imageio/TurboJpegImageReader.java [hunk 1]
```diff
@@ -116,6 +120,23 @@ public Iterator<ImageTypeSpecifier> getImageTypes(int imageIndex) {
...
+   * which is wrong, because it is under the user-specified threshold of 100.
+   * @param firstValue The first value to be checked
+   * @param secondValue The second value to be checked
+   * @param minValue The minimum value
+   * @return Integer of the closest value w.r.t. a given min value.
+   *         If both values are under the min value, min value will be returned
+   */
+  int getClosestValue(int firstValue, int secondValue, int minValue) {
+    Set<Integer> collectedValues = Stream.of(firstValue, secondValue).filter(val -> val >= minValue).collect(Collectors.toSet());
+    return collectedValues.isEmpty() ? minValue : Collections.min(collectedValues);
```
Comment (bitzl) [inline imageio-turbojpeg/src/main/java/de/digitalcollections/turbojpeg/imageio/TurboJpegImageReader.java:137]: Could this be written as `Math.max(Math.min(firstValue, secondValue), minValue)`?

**Review by bitzl — APPROVED** [review]

File: imageio-turbojpeg/src/main/java/de/digitalcollections/turbojpeg/imageio/TurboJpegImageReader.java (hunk 1 above)
Comment (stefan-it) [reply imageio-turbojpeg/src/main/java/de/digitalcollections/turbojpeg/imageio/TurboJpegImageReader.java:137]: This is e.g. not working with:

```java
int firstValue = 107;
int secondValue = 44;
int minValue = 100;
```

Original function would collect 107 which is right in this case.

Suggested function would return 100 which is not what we want here (100 should only be returned when `firstValue` and `secondValue` return smaller values than `minValue`.

Comment (jbaiter) [reply imageio-turbojpeg/src/main/java/de/digitalcollections/turbojpeg/imageio/TurboJpegImageReader.java:137]: I still think that using a Stream for this is probably overkill, this should do the same, but result in much faster execution:

```suggestion
    if (firstValue < minVal && secondValue < minVal) {
      return minVal;
    } else {
      return Math.min(firstVal, secondVal);
    }
```

Comment (jbaiter) [reply imageio-turbojpeg/src/main/java/de/digitalcollections/turbojpeg/imageio/TurboJpegImageReader.java:137]: how about this stream-less implementation?

```
if (firstVal < minVal && secondVal < minVal) {
  return minVal;
} else if (firstVal >= minVal) {
  return firstVal;
} else {
  return secondVal;
}
```

Should be much quicker!
```

Comment (stefan-it) [reply imageio-turbojpeg/src/main/java/de/digitalcollections/turbojpeg/imageio/TurboJpegImageReader.java]: Thanks!

**Review by jbaiter — APPROVED** [review]

**Pull request merged**